### ISSUE TITLE
Documentation updates

### ISSUE THREAD

**Issue by chadaustin** [libfuse/libfuse#333]
I'd like to submit some updates to libfuse's doxygen documentation, but don't have time right now. I'll briefly describe them here in case someone gets to it first.

`fuse_lowlevel_ops::lookup` should mention that it increases the inode's reference count and a corresponding `::forget` will arrive for that inode.

`fuse_lowlevel_ops::open` should mention that returning ENOSYS from open() will prevent release() calls from arriving as well.

`fuse_lowlevel_ops::release` says "For every open call there will be exactly one release call." which is not true under unmount.

`fuse_lowlevel_ops::releasedir` says "For every opendir call there will be exactly one releasedir call." which is not true under unmount.

`fuse_lowlevel_ops::readdir` says "The function does not have to report the '.' and '..' entries, but is allowed to do so." which is true, but this behavior is observable to userspace.

`fuse_lowlevel_ops::opendir` says "Filesystem may also implement stateless directory I/O and not store anything in fi->fh, though that makes it impossible to implement standard conforming directory stream operations in case the contents of the directory can change between opendir and releasedir." but that's not true - the 63 usable bits of off_t passed between readdir calls can be treated as seekdir/readdir cookies and index to the appropriate location in the directory stream, even under concurrent modification. For example, see https://github.com/facebookexperimental/eden/blob/ad044f2da05f2fad54515f950747e336c6add50f/eden/fs/inodes/TreeInode.cpp#L1700

**Comment by Nikratio** [comment]
Thanks for the suggestions, looking forward to the pull request :-).

Some comments:
- Using off_t as a cookie is stateful - you're saving state in the cookie
- Could you clarify what you mean with "not true under unmount"? If there are open files, you cannot unmount.
- If you add explicit lookup count related stuff to lookup(), we'll also need it to every other handler that affects lookup count - this is why the description is currently in just one place (forget)

**Comment by chadaustin** [comment]
Agreed wrt lookup - I will leave the documentation in forget. :)

You can `umount -fl $MOUNT` even with an open file handle, at least in my testing.

```
devvm835:~/source $ python3
>>> import os
>>> fd=os.open('.jfconfig', os.O_RDONLY)
>>> # ran `sudo umount -lf ~/source` here
>>> # verified the mount is no longer in the mount table
>>> os.read(fd, 10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 107] Transport endpoint is not connected
```

**Comment by Nikratio** [comment]
Well, using `--lazy` isn't actually doing a proper umount, it just makes the mount invisible. But it seems in your case it causes the filesystem to crash, so it seems something odd is going on. Is that one of the example filesystems? If so, please file a bug.

On the other hand, if you truly need `--force`, then all bets are off. That definitely does not count.

**Comment by chadaustin** [comment]
I just verified that, while `umount -f` fails on a FUSE mount that has an open file handle in it, the FUSE daemon exits, causing a connection error on read.

```
# run meson and ninja to build a libfuse project on a stock Ubuntu 18.04
# kernel 4.15.0-43-generic
$ PATH="$HOME/projects/libfuse/build/util:$PATH" ./hello -f ~/tmp/hello-mount

# in another terminal
$ cd tmp/hello-mount/
$ ls -lA
total 0
-r--r--r-- 1 root root 13 Jan  1  1970 hello
$ python3 -q
>>> import os
>>> os.open('hello', os.O_RDONLY)
3

# in another terminal
$ sudo umount -f ~/tmp/hello-mount
umount: /home/chadaustin/tmp/hello-mount: target is busy.

# in original terminal, confirm `hello` command has finished

# in Python prompt:
>>> os.read(3, 40)
OSError: [Errno 107] Transport endpoint is not connected
```

Is that what you'd expect?

**Comment by Nikratio** [comment]
TBH, I don't think we should care about the behavior when using `umount -f`. In that case, there's probably a number of other documented invariants that no longer apply. I'd be fine adding a general remark to that extent at the beginning at the docs. But description of individual functions should assume that the filesystem is unmounted regularly.

**Comment by chadaustin** [comment]
The reason I think it's worth clarifying `umount -f` is because, in reality (at least at Facebook) clean unmounts are quite rare. When a mount is running, it often has the buckd build daemon running in it, watchman, various other daemons, and they're all opening file handles. So we've found in practice that we need to force unmount most of the time.

In addition, the wording of the documentation makes it seem as if the filesystem would be correct to assume that every file handle has been released before it receives ENODEV from the fuse socket. But that's not actually true - the filesystem itself cannot observe whether a mount has been cleanly or force unmounted, and so it has to handle the possibility that all of the release() and releasedir() calls have not necessarily arrived. (Before we stopped handling open/release/opendir/releasedir, our filesystem made some incorrect assumptions here.)

I updated the PR to be more direct, merely pointing out that the filesystem should handle the possibility that release and releasedir may not always arrive before unmount and the filesystem should handle that.

Let me know what you think!
